File: src/net/azib/photos/XMLParser.java
package net.azib.photos;

import com.google.common.base.Joiner;

import javax.xml.stream.XMLEventReader;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.events.Attribute;
import javax.xml.stream.events.EndElement;
import javax.xml.stream.events.StartElement;
import javax.xml.stream.events.XMLEvent;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.util.Deque;
import java.util.Iterator;

import static com.google.common.collect.Lists.newLinkedList;
import static javax.xml.stream.XMLStreamConstants.*;

public class XMLParser<T> {
  private String characters = "";
  private final Deque<String> path = newLinkedList();
  private boolean rootFound;
  private final XMLListener<T> listener;

  public XMLParser(XMLListener<T> listener) {
    this.listener = listener;
  }

  public T parse(InputStream xml) {
    try {
      XMLEventReader reader = XMLInputFactory.newInstance().createXMLEventReader(xml);

      try {
        while (reader.hasNext()) {
          XMLEvent event = reader.nextEvent();
          switch (event.getEventType()) {
            case START_ELEMENT:
              startElement(event);
              break;
            case CHARACTERS:
              characters(event);
              break;
            case END_ELEMENT:
              endElement(event);
          }
        }
      }
      catch (XMLListener.StopParse e) {
        e.printStackTrace();
      }
      finally {
        reader.close();
      }
    }
    catch (XMLStreamException e) {
      throw new RuntimeException(e);
    }
    finally {
      closeQuietly(xml);
    }
    return listener.getResult();
  }

  public static void closeQuietly(Closeable closeable) {
    try {
      if (closeable != null) closeable.close();
    } catch (IOException ignore) {
    }
  }


  private void startElement(XMLEvent event) throws XMLListener.StopParse {
    characters = "";
    StartElement startElement = event.asStartElement();
    String name = startElement.getName().getLocalPart();
    if (!rootFound) {
      rootFound = true;
      listener.rootElement(name);
      parseAttributes(startElement, "");
      return;
    }
    path.addLast(name);
    String prefix = Joiner.on("/").join(path);
    listener.start(prefix);
    parseAttributes(startElement, prefix);
  }

  private void parseAttributes(StartElement element, String prefix) throws XMLListener.StopParse {
    for (Iterator<Attribute> attributes = element.getAttributes(); attributes.hasNext(); ) {
      Attribute attribute = attributes.next();
      listener.value(prefix + "@" + attribute.getName(), attribute.getValue());
    }
  }

  private void characters(XMLEvent event) {
    characters += event.asCharacters().getData();
  }

  private void endElement(XMLEvent event) throws XMLListener.StopParse {
    characters = characters.trim();
    if (!path.isEmpty()) {
      String p = Joiner.on("/").join(path);
      if (!characters.isEmpty()) {
        listener.value(p, characters);
        characters = "";
      }
      listener.end(p);
    }
    else {
      EndElement endElement = event.asEndElement();
      listener.rootElementEnd(endElement.getName().getLocalPart());
    }
    path.pollLast();
  }
}

File: src/net/azib/photos/XMLParser.kt
package net.azib.photos

import java.io.InputStream
import java.util.*
import javax.xml.stream.XMLInputFactory
import javax.xml.stream.XMLStreamConstants.*
import javax.xml.stream.events.Attribute
import javax.xml.stream.events.StartElement
import javax.xml.stream.events.XMLEvent

class XMLParser<T>(private val listener: XMLListener<T>) {
  private var characters = ""
  private val path = LinkedList<String>()
  private var rootFound = false

  fun parse(xml: InputStream): T = xml.use {
    val reader = XMLInputFactory.newInstance().createXMLEventReader(xml)
    try {
      while (reader.hasNext()) {
        val event = reader.nextEvent()
        when (event.eventType) {
          START_ELEMENT -> startElement(event)
          CHARACTERS -> characters(event)
          END_ELEMENT -> endElement(event)
        }
      }
    }
    catch (e: XMLListener.StopParse) {
    }
    finally {
      reader.close()
    }
    return listener.result
  }


  private fun startElement(event: XMLEvent) {
    characters = ""
    val startElement = event.asStartElement()
    val name = startElement.name.localPart
    if (!rootFound) {
      rootFound = true
      listener.rootElement(name)
      parseAttributes(startElement, "")
      return
    }
    path.addLast(name)
    val prefix = path.joinToString("/")
    listener.start(prefix)
    parseAttributes(startElement, prefix)
  }

  private fun parseAttributes(element: StartElement, prefix: String) {
    @Suppress("UNCHECKED_CAST")
    val attributes = element.attributes as Iterator<Attribute>
    while (attributes.hasNext()) {
      val attribute = attributes.next()
      listener.value(prefix + "@" + attribute.name, attribute.value)
    }
  }

  private fun characters(event: XMLEvent) {
    characters += event.asCharacters().data
  }

  private fun endElement(event: XMLEvent) {
    characters = characters.trim()
    if (!path.isEmpty()) {
      val p = path.joinToString("/")
      if (!characters.isEmpty()) {
        listener.value(p, characters)
        characters = ""
      }
      listener.end(p)
    }
    else {
      val endElement = event.asEndElement()
      listener.rootElementEnd(endElement.name.localPart)
    }
    path.pollLast()
  }
}
